### PR TITLE
Fix list enumeration

### PR DESCRIPTION
**Pull request by nxtn**
# PR Summary

Fixes list enumeration and removal.

## PR Context

A collection shall not be modified during enumeration.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
- **User-facing changes**
    - [x] Not Applicable
- **Testing - New and feature**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 1]
```diff
@@ -1106,12 +1106,13 @@ private bool HaveAllDependentServicesStopped(ICollection<ServiceController> depe
         /// <param name="services">A list of services.</param>
         internal void RemoveNotStoppedServices(List<ServiceController> services)
         {
-            foreach (ServiceController service in services)
+            for (int i = services.Count - 1; i >= 0; i--)
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Is there a reason you're enumerating from the end to the beginning here?

Also, I don't see a real need to use a for loop here. For loops are nice, but tend to create confusing and less readable code on the whole.

Perhaps a better solution would be to attempt to stop each service, then filter out the stopped services with one of the List filter methods.

Comment (NextTurn) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: > Is there a reason you're enumerating from the end to the beginning here?

Enumerating from the end tends to move fewer items.

> Also, I don't see a real need to use a for loop here. For loops are nice, but tend to create confusing and less readable code on the whole.

Sounds good. Updated to use `RemoveAll` (and `TrueForAll` for another method BTW).

> Perhaps an alternative solution would be to attempt to stop each service

The services have been requested to stop before this filter method.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 2]
```diff
@@ -1087,17 +1087,9 @@ internal List<ServiceController> DoStopService(ServiceController serviceControll
         /// True if all dependent services are stopped
         /// False if not all dependent services are stopped
         /// </returns>
-        private bool HaveAllDependentServicesStopped(ICollection<ServiceController> dependentServices)
+        private bool HaveAllDependentServicesStopped(ServiceController[] dependentServices)
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: I'm not sure what the possible impact of changing the method signature here would be, but given it's a private method anyway, it's probably safe to do as long as the code calling the method is still OK to call it with this change.

**Review by vexx32 — APPROVED** [review]
Looks good to me overall! One minor thing you might want to double check just to be sure it won't cause any issues, but it doesn't seem like it should break anything as far as I can see.

Thanks for sorting this out! 😊 

If there's a scenario you know was previously broken, it might be a good idea to add a test to ensure that scenario is completely fixed with this change and does not regress later on. 🙂

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 2 above)
Comment (NextTurn) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: This method is only called from one place, hence I believe it's separated to provide better readability.

**Comment by NextTurn** [comment]
> If there's a scenario you know was previously broken, it might be a good idea to add a test to ensure that scenario is completely fixed with this change and does not regress later on.

It makes sense to cover these code paths. The tests will require a custom service, which refuses to accept the stop command during testing but accepts it when uninstalled...

I don't want to tackle that at this moment. :(

**Comment by vexx32** [comment]
That does sound rather complicated, indeed. 

The fix seems logically sound, though, so unless there's a more straightforward route to testing it, that might have to come later. 🙂

**Comment by iSazonov** [comment]
@NextTurn Please recover PR template to original and add a reference to an issue with a discussion.

**Comment by NextTurn** [comment]
> Please recover PR template to original and add a reference to an issue with a discussion.

Opened #11858

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 3]
```diff
@@ -1106,14 +1098,9 @@ private bool HaveAllDependentServicesStopped(ICollection<ServiceController> depe
...
         {
-            foreach (ServiceController service in services)
-            {
-                if (service.Status != ServiceControllerStatus.Stopped &&
-                    service.Status != ServiceControllerStatus.StopPending)
-                {
-                    services.Remove(service);
-                }
-            }
+            services.RemoveAll(service =>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Please add a comment why we have to use LINQ.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 3 above)
Comment (NextTurn) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: This is not LINQ. It's a method of `List<T>`.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: It is just better. Please add a comment.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 4]
```diff
@@ -1098,6 +1098,8 @@ private bool HaveAllDependentServicesStopped(ServiceController[] dependentServic
         /// <param name="services">A list of services.</param>
         internal void RemoveNotStoppedServices(List<ServiceController> services)
         {
+            // You shall not modify a collection during enumeration.
+            // https://stackoverflow.com/questions/1582285/how-to-remove-elements-from-a-generic-list-while-iterating-over-it
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: No need to reference StackOverflow. A simple comment is enough that we cannot use an enumerator.

**Comment by iSazonov** [comment]
@TravisEz13 @daxian-dbw The PR is ready to merge.

**Review by daxian-dbw — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@NextTurn Thank you for your contribution!

**Pull request merged**